src/parser.rs

```diff
@@ -0,0 +1,367 @@
+    name_token_is_none: bool,
+  ) -> Option<Token> {
+    // TODO: do not mut, instead return
+    let mut token = self.try_consume_token(TokenType::Regexp);
+    if name_token_is_none && token.is_none() {
+      token = self.try_consume_token(TokenType::Asterisk);
+    }
+    token
+  }
+
+  // Ref: https://wicg.github.io/urlpattern/#try-to-consume-a-modifier-token
+  // TODO: inline?
+  fn try_consume_modifier_token(&mut self) -> Option<Token> {
+    // TODO: use .or()
+    let token = self.try_consume_token(TokenType::OtherModifier);
+    if token.is_some() {
+      token
+    } else {
+      self.try_consume_token(TokenType::Asterisk)
+    }
+  }
+
+  // Ref: https://wicg.github.io/urlpattern/#maybe-add-a-part-from-the-pending-fixed-value
+  // TODO: inline?
+  fn maybe_add_part_from_pending_fixed_value(
+    &mut self,
+  ) -> Result<(), ParseError> {
+    if self.pending_fixed_value.is_empty() {
+      return Ok(());
+    }
+    let encoded_value = (self.encoding_callback)(&self.pending_fixed_value)?;
+    self.pending_fixed_value = String::new();
+    self.part_list.push(Part::new(
+      PartType::FixedText,
+      encoded_value,
+      PartModifier::None,
+    ));
+
+    Ok(())
+  }
+
+  // Ref: https://wicg.github.io/urlpattern/#add-a-part
+  fn add_part(
+    &mut self,
+    prefix: &str,
+    name_token: Option<Token>,
+    regexp_or_wildcard_token: Option<Token>,
+    suffix: &str,
+    modifier_token: Option<Token>,
+  ) -> Result<(), ParseError> {
+    let mut modifier = PartModifier::None;
+    if let Some(modifier_token) = modifier_token {
+      modifier = match modifier_token.value.as_ref() {
+        "?" => PartModifier::Optional,
+        "*" => PartModifier::ZeroOrMore,
+        "+" => PartModifier::OneOrMore,
+        _ => unreachable!(),
+      };
+    }
+    if name_token.is_none()
+      && regexp_or_wildcard_token.is_none()
+      && modifier == PartModifier::None
+    {
+      self.pending_fixed_value.push_str(prefix);
+      return Ok(());
+    }
+    self.maybe_add_part_from_pending_fixed_value()?;
+    if name_token.is_none() && regexp_or_wildcard_token.is_none() {
+      assert!(suffix.is_empty());
+      if prefix.is_empty() {
+        return Ok(());
+      }
+      let encoded_value = (self.encoding_callback)(prefix)?;
+      self.part_list.push(Part::new(
+        PartType::FixedText,
+        encoded_value,
+        modifier,
+      ));
+      return Ok(());
+    }
+
+    let mut regexp_value = match &regexp_or_wildcard_token {
+      None => self.segment_wildcard_regexp.to_owned(),
+      Some(regexp_or_wildcard_token) => {
+        if regexp_or_wildcard_token.kind == TokenType::Asterisk {
+          FULL_WILDCARD_REGEXP_VALUE.to_string()
+        } else {
+          regexp_or_wildcard_token.value.to_owned()
+        }
+      }
+    };
+
+    let mut kind = PartType::Regexp;
+    if regexp_value == self.segment_wildcard_regexp {
+      kind = PartType::SegmentWildcard;
+      regexp_value = String::new();
```

Comment:
Would [`String.clear()`](https://doc.rust-lang.org/std/string/struct.String.html#method.clear) be better for resetting the contents here and in other places like this?  Although, I'm not sure if the part of it not resetting the capacity could be considered bad.

```suggestion
      regexp_value.clear();
```